lib/lndk.js

```diff
@@ -0,0 +1,249 @@
+      response_invoice_timeout: timeout
+    }, async (error, response) => {
+      if (error) return reject(error)
+      const bech32invoice = 'lni1' + bech32b12.encode(Buffer.from(response.invoice_hex_str, 'hex'))
+
+      // sanity check
+      const parsedInvoice = await parseBolt12Request({ lnd, request: bech32invoice })
+      if (
+        !parsedInvoice ||
+        toPositiveNumber(parsedInvoice.mtokens) !== toPositiveNumber(msats) ||
+        toPositiveNumber(parsedInvoice.tokens) !== toPositiveNumber(msatsToSats(msats))
+      ) {
+        return reject(new Error('invalid invoice response'))
+      }
+      resolve(bech32invoice)
+    })
+  })
+}
+
+export async function payViaBolt12PaymentRequest ({
+  lnd,
+  request: invoice_hex_str,
+  max_fee,
+  max_fee_mtokens
+}) {
+  const lndk = lnd?.lndk
+  if (!lndk) throw new Error('lndk not installed, please use installLNDK')
+
+  const parsedInvoice = await parseBolt12Request({ lnd, request: invoice_hex_str })
+
+  if (!max_fee_mtokens && max_fee) {
+    max_fee_mtokens = toPositiveNumber(satsToMsats(max_fee))
+  }
+
+  return new Promise((resolve, reject) => {
+    lndk.PayInvoice({
+      invoice: parsedInvoice.payment,
+      // expects msats amount: https://github.com/lndk-org/lndk/blob/bce93885f5fc97f3ceb15dc470117e10061de018/src/lib.rs#L403
+      amount: toPositiveNumber(parsedInvoice.mtokens),
+      max_fee: toPositiveNumber(max_fee_mtokens)
+    }, (error, response) => {
+      if (error) {
+        return reject(error)
+      }
+      resolve({
+        secret: response.payment_preimage
+      })
+    })
+  })
+}
+
+export async function estimateBolt12RouteFee ({ lnd, destination, tokens, mtokens, request, timeout }) {
+  const lndk = lnd?.lndk
+  if (!lndk) throw new Error('lndk not installed, please use installLNDK')
+  const parsedInvoice = request ? await parseBolt12Request({ lnd, request }) : {}
+
+  if (!tokens && mtokens) tokens = toPositiveNumber(msatsToSats(mtokens))
+  tokens ??= toPositiveNumber(parsedInvoice.tokens)
+  destination ??= parsedInvoice.destination
+
+  if (!destination) throw new Error('no destination provided')
+  if (!tokens) throw new Error('no tokens provided')
+
+  return await new Promise((resolve, reject) => {
+    lnd.router.estimateRouteFee({
+      dest: Buffer.from(destination, 'hex'),
+      amt_sat: tokens,
+      timeout
+    }, (err, res) => {
+      if (err) {
+        if (res?.failure_reason) {
+          reject(new Error(`unable to estimate route: ${res.failure_reason}`))
+        } else {
+          reject(err)
+        }
+        return
+      }
+
+      if (res.routing_fee_msat < 0 || res.time_lock_delay <= 0) {
+        reject(new Error('unable to estimate route, excessive values: ' + JSON.stringify(res)))
+        return
+      }
+
+      resolve({
+        routingFeeMsat: toPositiveNumber(res.routing_fee_msat),
+        timeLockDelay: toPositiveNumber(res.time_lock_delay)
+      })
+    })
+  })
+}
```

Comment:
I think this is the same as `estimateRouteFee` for bolt11 except it needs to parse the bolt12 request before. This should afterwards call `estimateRouteFee` imo.